sdk/eventgrid/azure-eventgrid/tests/eventgrid_preparer.py

```diff
@@ -0,0 +1,98 @@
+import functools
+import hashlib
+import os
+from collections import namedtuple
+
+from azure.mgmt.eventgrid import EventGridManagementClient
+from azure.mgmt.eventgrid.models import Topic, InputSchema, JsonInputSchemaMapping, JsonField, JsonFieldWithDefault
+from azure_devtools.scenario_tests.exceptions import AzureTestError
+
+from devtools_testutils import (
+    ResourceGroupPreparer, AzureMgmtPreparer, FakeResource
+)
+
+from devtools_testutils.resource_testcase import RESOURCE_GROUP_PARAM
+
+EVENTGRID_TOPIC_PARAM = 'eventgrid_topic'
+EVENTGRID_TOPIC_LOCATION = 'westus'
+CLOUD_EVENT_SCHEMA = InputSchema.cloud_event_schema_v1_0
+CUSTOM_EVENT_SCHEMA = InputSchema.custom_event_schema
+ID_JSON_FIELD = JsonField(source_field='customId')
+TOPIC_JSON_FIELD = JsonField(source_field='customTopic')
+EVENT_TIME_JSON_FIELD = JsonField(source_field='customEventTime')
+EVENT_TYPE_JSON_FIELD_WITH_DEFAULT = JsonFieldWithDefault(source_field='customEventType', default_value='')
+SUBJECT_JSON_FIELD_WITH_DEFAULT = JsonFieldWithDefault(source_field='customSubject', default_value='')
+DATA_VERSION_JSON_FIELD_WITH_DEFAULT = JsonFieldWithDefault(source_field='customDataVersion', default_value='')
+CUSTOM_JSON_INPUT_SCHEMA_MAPPING = JsonInputSchemaMapping(id=ID_JSON_FIELD, topic=TOPIC_JSON_FIELD, event_time=EVENT_TIME_JSON_FIELD, event_type=EVENT_TYPE_JSON_FIELD_WITH_DEFAULT, subject=SUBJECT_JSON_FIELD_WITH_DEFAULT, data_version=DATA_VERSION_JSON_FIELD_WITH_DEFAULT)
+
+
+class EventGridTopicPreparer(AzureMgmtPreparer):
+    def __init__(self,
+                 name_prefix='',
+                 use_cache=False,
+                 parameter_location=EVENTGRID_TOPIC_LOCATION,
+                 parameter_name=EVENTGRID_TOPIC_PARAM,
+                 resource_group_parameter_name=RESOURCE_GROUP_PARAM,
+                 disable_recording=True, playback_fake_resource=None,
+                 client_kwargs=None, random_name_enabled=True):
+        super(EventGridTopicPreparer, self).__init__(name_prefix, random_name_length=24,
+                                                     random_name_enabled=random_name_enabled,
+                                                     disable_recording=disable_recording,
+                                                     playback_fake_resource=playback_fake_resource,
+                                                     client_kwargs=client_kwargs)
+        self.resource_group_parameter_name = resource_group_parameter_name
+        self.parameter_name = parameter_name
+        self.parameter_location = parameter_location
+        self.name_prefix = name_prefix
+        if random_name_enabled:
+            self.resource_moniker = self.name_prefix + "egtopic"
+        
+        self.set_cache(use_cache, name_prefix)
```

Comment:
name_prefix isn't necessary for caching since it doesn't change the nature of the resource.  (SKU for instance, or region, would change the resource)

Reply:
Still seems to be present?

Reply:
From our OOB conversation; there's buggy interplay between the cache and name_prefix.  I've taken an action item on myself to look into that, for now this is fine.